worker/src/features/evaluation/evalService.ts

```diff
@@ -87,225 +90,110 @@ export const createTraceEvalJobs = async ({
+            )
+          : await kyselyPrisma.$kysely
+              .selectFrom("observations")
+              .select("id")
+              .where("project_id", "=", event.projectId)
+              .where("id", "=", event.observationId)
+              .executeTakeFirst();
+
+      if (!observationExists) {
+        logger.warn(
+          `Observation ${event.observationId} not found, retrying dataset eval later`,
+        );
+        throw new Error(
+          "Observation not found. Rejecting job to use retry-attempts.",
+        );
+      }
+    }
 
     const existingJob = await kyselyPrisma.$kysely
       .selectFrom("job_executions")
       .select("id")
       .where("project_id", "=", event.projectId)
       .where("job_configuration_id", "=", config.id)
-      .where("job_input_dataset_item_id", "=", event.datasetItemId)
       .where("job_input_trace_id", "=", event.traceId)
+      .where(
+        "job_input_dataset_item_id",
+        isDataSetItemRun ? "=" : "is",
+        isDataSetItemRun ? event.datasetItemId : null,
+      )
       .where(
         "job_input_observation_id",
-        event.observationId ? "=" : "is",
-        event.observationId ?? null,
+        "observationId" in event && event.observationId ? "=" : "is",
+        "observationId" in event && event.observationId
+          ? event.observationId
+          : null,
       )
       .execute();
 
-    // if we matched a dataset item, we might want to create a job
-    if (datasetItems.length > 0) {
-      // check if trace and observation exist, otherwise retry
-      // Verify trace exists before proceeding
-      const traceExists = await kyselyPrisma.$kysely
-        .selectFrom("traces")
-        .select("id")
-        .where("project_id", "=", event.projectId)
-        .where("id", "=", event.traceId)
-        .executeTakeFirst();
-
-      if (!traceExists) {
-        logger.info(
-          `Trace ${event.traceId} not found, retrying dataset eval later`,
-        );
-        throw new Error("Trace not found - initiate retry");
-      }
-
-      if (event.observationId) {
-        const observationExists = await kyselyPrisma.$kysely
-          .selectFrom("observations")
-          .select("id")
-          .where("project_id", "=", event.projectId)
-          .where("id", "=", event.observationId)
-          .executeTakeFirst();
-
-        if (!observationExists) {
-          logger.info(
-            `Observation ${event.observationId} not found, retrying dataset eval later`,
-          );
-          throw new Error("Observation not found - initiate retry");
-        }
-      }
-
-      logger.debug(
-        `Eval job for config ${config.id} matched dataset run item ids ${JSON.stringify(datasetItems.map((d) => d.id))}`,
-      );
-
+    // If we matched a trace and datasetitem, we might want to create a job
+    // If this is only run for a trace, datasetItemExists always default to true.
+    if (traceExists && datasetItemExists) {
```

Comment:
Nit: as the code is quite complex leading up to this, I think we could be clearer on this comment. The second part is helpful, I would lead with that. I would also provide more context in the comment on the flow of this code, as someone must understand: 
- that if we are dealing with an event from the trace upsert queue we don't check if dataset items exist (irrelevant), and that defaults to true 
- that if we are dealing with an event from the dataset items queue, we require the corresponding trace to be available for the eval job to be successful